// angular-user-idle/esm2020/lib/angular-user-idle.module.mjs
import { NgModule } from '@angular/core';
import { UserIdleConfig } from './angular-user-idle.config';
import * as i0 from "@angular/core";
export class UserIdleModule {
    static forRoot(config) {
        return {
            ngModule: UserIdleModule,
            providers: [{ provide: UserIdleConfig, useValue: config }],
        };
    }
}
UserIdleModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.11", ngImport: i0, type: UserIdleModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
UserIdleModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.11", ngImport: i0, type: UserIdleModule });
UserIdleModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.11", ngImport: i0, type: UserIdleModule, imports: [[]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.11", ngImport: i0, type: UserIdleModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [],
                }]
        }] });
