// @coreui/angular/esm2020/lib/modal/modal-content/modal-content.component.mjs
import { Component, HostBinding, HostListener } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "@angular/cdk/a11y";
export class ModalContentComponent {
    constructor(hostElement, interactivityChecker) {
        this.hostElement = hostElement;
        this.interactivityChecker = interactivityChecker;
    }
    get hostClasses() {
        return {
            'modal-content': true
        };
    }
    onClickHandler($event) {
        $event.stopPropagation();
        const targetElement = $event.target;
        return this.hostElement?.nativeElement?.contains(targetElement) &&
            this.interactivityChecker.isFocusable(targetElement);
    }
}
ModalContentComponent.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.0.5", ngImport: i0, type: ModalContentComponent, deps: [{ token: i0.ElementRef }, { token: i1.InteractivityChecker }], target: i0.ɵɵFactoryTarget.Component });
ModalContentComponent.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "14.0.5", type: ModalContentComponent, selector: "c-modal-content", host: { listeners: { "click": "onClickHandler($event)" }, properties: { "class": "this.hostClasses" } }, ngImport: i0, template: '<ng-content></ng-content>', isInline: true });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.0.5", ngImport: i0, type: ModalContentComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'c-modal-content',
                    template: '<ng-content></ng-content>'
                }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef }, { type: i1.InteractivityChecker }]; }, propDecorators: { hostClasses: [{
                type: HostBinding,
                args: ['class']
            }], onClickHandler: [{
                type: HostListener,
                args: ['click', ['$event']]
            }] } });
